Board Layout Changes and Legal Moves Fix

MfonUdoh opened the pull request:
Hi, 

- I edited the draw_board function a bit to make the board print out easier to read ie. gridlines and postion labels.

- Modified the HumanPlayer class to check the player moves are legal before allowing the move to be executed and allowed legal capture moves.

MfonUdoh committed:
Update label.py

MfonUdoh committed:
Update game_state.py

Modified the draw_board function to print the game board in a nice grid that is easier to read and makes moving pieces easier to understand

MfonUdoh committed:
Update hexapawn.py

Shows completed board at the end of the game

MfonUdoh committed:
Fixed Legal Player Moves Check

Originally the player was able to cheat and make illegal moves (moves to any blank space) and would not allow legal attack moves (diagonal moves to capture enemy pieces). I have modified this code so it prompts for the row and column individually and then checks that the requsted move is available using the get_available_moves function, then returns the move if legal.

Hydrotoast commented on game_state.py:
```diff
@@ -146,10 +146,23 @@ def get_pawn(self, row, col, color):
-        for i in range(3):
-            print(self.board[i])
-        print()
+    def draw_board(self):        
```
Important. This applies to all changes in the patch. To maintain consistency with the rest of the code, indent with 4 spaces in this method.

MfonUdoh replied:
I have aimed to use this indentation everywhere, please highlight anywhere you've spotted I haven't

Hydrotoast replied:
Thanks for handling this! I know that it can be tedious to keep up the style. I see the indentation is still at 2 spaces from lines 157 to 164, but I imagine you will resolve after looking into cleaning up that region.

MfonUdoh replied:
Okay I see what you mean, it looks very out of place with only 2 spaces, at the time I wrote that section of code I was using a more basic IDE, since then I've switched to Visual Studio that follows the indentation formatting automatically for you. 
For my knowledge, do you know of anyway to check that the indentation is consistent automatically or is it a case of inspection by eye?

Hydrotoast replied:
Yes, the style guide that the majority of Python libraries follow is known as [PEP 8](https://pep8.org/). If you enjoy writing Python code, I suggest scanning through the examples occassionally to help remind yourself of what good style looks like.

To automate the process for consistent style, the tools that do this are known as "linters". See [Pylint](https://www.pylint.org/) as an example.

Hydrotoast commented:
Hey @MfonUdoh,

Thanks for the patches to this project! I have not worked on this project in several years, so I may be slower to review this code than usual. Hopefully, you can be patient with me. :)

Based on your summary, I believe the objective for this patch is to improve the friendliness to users by: (1) printing the board in a readable representation and (2) recovering from simple user mistakes.

I will scan through the patch and perhaps leave a few comments on some diffs. If I believe that comment is important to address, I will formalize it as a request for changes. Please address requests for changes before we proceed to merge, but feel free to discuss the request or contest my requests. Any other general comment does not need to be addressed, but I may leave some suggestions as general comments as well.

Once the reviews have been addressed, we can proceed to merge your patches.

Note that I may include a `LICENSE` beforehand to avoid any subtle legal issues with contributions. Perhaps BSD 3-clause or Apache V2. This `LICENSE` file should not affect conflict  with a merge. However, if you have general concerns with this, I can discuss it further.

Hydrotoast commented on game_state.py:
```diff
@@ -146,10 +146,23 @@ def get_pawn(self, row, col, color):
+      print(' ', end=' ')
+      blabels = [str(i) for i in range(3)]
+      for ele in blabels:
+        print("|",ele, end=" ")
```
Important. This applies to all changes in the patch. To maintain consistent argument separation with the rest of the code, add a space before the start of subsequent arguments, e.g. `fn(arg1, arg2, ..., arg3)`.

Hydrotoast commented on game_state.py:
```diff
@@ -146,10 +146,23 @@ def get_pawn(self, row, col, color):
+      for ele in blabels:
+        print("|",ele, end=" ")
+      print('|')
+      print('-'*15)
```
Important. This applies to all changes in the patch. To maintain consistent infix operations with the rest of the code, add a space before and after each operation, e.g. `arg1 * arg2`.

Hydrotoast commented on player.py:
```diff
@@ -32,19 +32,26 @@ def make_move(self, state):
+            row = eval(input('Select a row: '))
+            col = eval(input('Select a column: '))
+            for moveset in state.get_available_moves():
+                if repr(Move(pawn, Position(row, col))) == repr(moveset):
```
Bug? I believe that this proposition is always false because: (1) the left-hand side represents a move `Move(...)` and (2) the right-hand side represents a _list_ of moves `[...]`. So prefix and suffix should never match.

I believe the intent of this for-loop is to: (1) validate the user move and (2) construct the position. Perhaps we can make this more concise as follows.
```python
final_pos = Position(row, col)
valid = Move(pawn, final_pos)) in state.get_available_moves()
```

Thoughts?

MfonUdoh replied:
@Hydrotoast  This may stem from my confusing labeling of the 'for'-loop:

get_available_moves ; "gives a **list** of all the pairs of original and final positions that are legally allowed for the player".
I thought of these pairs of original and final positions as sets of moves so I named each element in the 'for'-loop, 'moveset'.

Hence, 
moveset; "A single pair of original and final positions that constructs a move".

So if the player selects for example: B at (2, 2) (1, 2)
the loop will confirm that it is a match for one of the elements (moveset) of the get_available_moves list

I can change the elements to something more clear like 'movepairs' for example

Hydrotoast replied:
You are correct. :)

I had misinterpreted the `moveset` as a list without running the code. I like the idea that `movepairs` would avoid ambiguity.

I also think that we could also cut down on several levels of indentation and simplify the logic by using the previous proposition (or some variant):

```python
final_pos = Position(row, col)
valid = Move(pawn, final_pos)) in state.get_available_moves()
```

Can you think of any other ways to simplify this code?

MfonUdoh replied:
Yes, this works wonderfully!
I originally tried to use something along these lines but couldn't get it to work until I used repr() but at that point I had already overcomplicated the process with the for-loop. Yes, this is much better.

Hydrotoast commented on player.py:
```diff
@@ -32,19 +32,26 @@ def make_move(self, state):
+            if (row in range(3)) and (col in range(3)):
+                pawn = state.get_pawn(row, col, self.color)
+                if pawn is not None:
+                    valid = True
```
Although I enjoy the idea of merging these two input loops, I think require an extra step from the user if the row and column selections are invalid. Consider when the `row` and `col` do not refer to a valid pawn. Then, the user would proceed to provide a final position even though the original pawn position is invalid!

Can we avoid this scenario?

MfonUdoh replied:
@Hydrotoast  It is true that the first loop would allow the user to continue to the second loop without a valid original position, however the second loop checks **both** the original position and the proposed final position using your 'get_available_moves' function. 
Therefore, even if the player doesn't select a valid original position, it will still be rejected at the second loop, only moves out of the legal moveset list will be allowed. 
I do agree that this does make the first loop kind of redundant as a 'check' however I left it in operation in order to grab the 'pawn' variable.

Hydrotoast replied:
I am okay with this resolution, but I do see one place we can simplify the code in lines 42 to 43: the conditional statement can be merged into one line.

```python
valid = pawn is not None
```

MfonUdoh replied:
Yes this is a good simplification.
Also I propose; like I suggested earlier the first loop is redundant as a check so it would be a good idea to wrap the second loop in an 'if' so we can actually use this loop as a check. 
`if valid == True:`
This way, if the user does not select a valid position, it will not bother asking for a final position it will skip this and ask the user to try again as soon as it knows the move will not work. 
If you think this is a good idea I will push the commit, if you think it is unnecessary and adds too much complication then I think it's best to remove the validation from the first loop since it will be checked in the second loop regardless.

Hydrotoast replied:
I like your idea. :)

Hydrotoast commented on player.py:
```diff
@@ -32,19 +32,26 @@ def make_move(self, state):
+            print('Begin by selecting a pawn in a row and column')
+            row = eval(input('Select a row: '))
+            col = eval(input('Select a column: '))
+            if (row in range(3)) and (col in range(3)):
```
Although the range is small, the `range(...)` function allocates memory, which is a slow operation overall. Maybe we can simplify with interval comparisons?

```python
if 1 <= row <= 3 and 1 <= col <= 3:
```

MfonUdoh commented:
@Hydrotoast that sounds great 👍 ! 
Don't worry about taking your time with the review, I'm aware that this project hasn't been touched in a few years, so I'm happy even to get a response. 

I stubbled upon your project after watching this [YouTube video](https://youtu.be/sw7UAZNgGg8) about Hexapawn and ML, I figured going through your code to learn how it works would be a good learning experience and would give me the opportunity to help along the way. So I look forward to addressing any comments that you may have. :)

Hydrotoast commented on game_state.py:
```diff
@@ -146,10 +146,23 @@ def get_pawn(self, row, col, color):
+    def draw_board(self):        
+      print()
+      print(' ', end=' ')
+      blabels = [str(i) for i in range(3)]
```
I suspect that many of these loops can be replaced with the `join` method on strings. For example, consider joining a list of integers with a vertical pipe:

```python
' | '.join([1, 2, 3]) == '1 | 2 | 3'
```

Can we simplify the logic in this function with this method?

MfonUdoh replied:
Yes, you are correct this method is a bit overcomplicated - I orginially used this to allow the board size to be changed dynamically in the future however I see this is unlikely to be required - I have replaced with a join of a simple list of strings.
`blabels = ['0', '1', '2']`
`print('|', ' | '.join(blabels), end=' ')`

Hydrotoast reviewed (changes requested):
Overall, I like the intent behind the patch. But I have a few requests for changes to consider some cases of unfriendly behavior and to maintain the overall quality of code in the project.

The requests can be categorized into two groups: maintaining consistent style and simplifying logic. First, the consistent style changes include requests such as indentation and spaces around operators. Second, the the simplifications intend for to maintain system performance, enable composable behavior, and provide readable source code for future contributors. :)

I may have further comments as we progress through this. Let me know if you have any questions, comments, or concerns about these requests.

Hydrotoast commented on hexapawn.py:
```diff
@@ -47,6 +47,7 @@ def play_game():
         print('Winner is RED player')
     else:
         print('Winner is BLUE player')
+    state.draw_board()
```
I like the idea of drawing the end state of a board after the end of a game! This seems like behavior that should been available from the beginning. :)

MfonUdoh committed:
Update game_state.py

Response to suggested logic simplification

MfonUdoh committed:
Update game_state.py

MfonUdoh committed:
Update player.py

Response to reduce memory allocation

Hydrotoast committed:
Refactor 'draw_board'

Hydrotoast reviewed (changes requested):
Thanks for quickly addressing the issues for the first review!

I have submitted a patch to your repository for the `draw_board` behavior to add some documentation on the new format and also refactor to keep the logic simple. I have done my best to retain the original logic, so if it deviates from what you have intended, feel free to overwrite my changes or discuss.

I had not seen the formatting before submitting the patch to your fork. Great job!

MfonUdoh committed:
Update player.py

Simplification of the player move check logic and elimated player position check redundancy

MfonUdoh reviewed (commented):
This is great work, the code looks much cleaner and simpler :+1:

MfonUdoh commented:
@Hydrotoast,
I've responded to all your comments and made some changes. 
Thanks for continuing to provide feedback, I very much appreciate your patience as I am new to Python, it has been a pleasure working with you on this project and I feel I am learning a lot about producing high quality code.

Hydrotoast commented:
Approved. Thanks for the contributions @MfonUdoh !

I have merged your changes and they are now officially part of the project. I appreciate the diligence to work through the reviews. Congratulations on being the first contributor!

Note. I have added a [BSD 3-clause](https://opensource.org/licenses/BSD-3-Clause) license, which is common among the scientific community, e.g. numpy and pandas. It is a simple open-source license intended to protect the contributions and the contributors.

Finally, if you would like the complete engineering experience, I recommend considering the following optional tasks after implementing a change to a project:

- [ ] Update the `README` to reflect your changes
- [ ] Update the `README` to help future contributors like yourself get started (especially if you found anything difficult)
- [ ] Add comments for docstrings for readers of your code. See [PEP 257](https://www.python.org/dev/peps/pep-0257/) on the style guide for it.
- [ ] Compare your style with a standard style, e.g. [Google Python Style](http://google.github.io/styleguide/pyguide.html).

Overall, don't feel compelled to complete these as formal tasks. Use your best judgement to decide which problems are worth solving. Generally, I think some of the most significant impact starts from thinking about how you can help someone else get started and work through the code. :)

If you would like to help maintain this project, perhaps a more advanced task for the project to automate its own maintenance:

- [ ] Setup continuous integration with [Travis CI](https://docs.travis-ci.com/user/languages/python/) to automatically test any changes

There are always projects out there in need of contributions. Great work so far!

MfonUdoh commented:
Okay, thank you for the Approval and Merge @Hydrotoast, 
I will look into updating the README and the others you suggested :grin:
